File: PythonSolutions/Neetcode150/graph_problems/417_pacific-atlantic-water-flow.py
from typing import List
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        self.ROWS, self.COLS = len(heights), len(heights[0])
        self.DIRS = [(0, 1), (1, 0), (-1, 0), (0, -1)]
        self.heights = heights
        
        pacific: set[tuple[int, int]] = set()
        atlantic: set[tuple[int, int]] = set()
        
        # Multi-source DFS below: Edges are adjacent to ocean so these cells can already go to that specific ocean

        # DFS through upper pacific and lower atlantic oceans
        for col in range(self.COLS):
            self.dfs(0, col, pacific)
            self.dfs(self.ROWS - 1, col, atlantic)

        # DFS through left pacific and right atlantic oceans
        for row in range(self.ROWS):
            self.dfs(row, 0, pacific)
            self.dfs(row, self.COLS - 1, atlantic)

        res = []
        # Intersection between pacific and atlantic oceans to see which cell can leak to both oceans
        for row in range(self.ROWS):
            for col in range(self.COLS):
                if (row, col) in pacific and (row, col) in atlantic:
                    res.append([row, col])
        
        return res
    
    def dfs(self, row, col, ocean):
        # Base case 1: We have already traversed here previously in another DFS so skip
            # We do not need to do another DFS in a traversed cell since we already know at that point if water can flow
        if (row, col) in ocean:
            return
        
        # Mark that the water can flow to the ocean
        ocean.add((row, col))
        
        for dx, dy in self.DIRS:
            nr, nc = row + dx, col + dy

            if nr < 0 or nc < 0 or nr > self.ROWS - 1 or nc > self.COLS - 1:
                continue
            
            # Only DFS the heights greater than current cell (we are following opposite of rain flow from cell)
            if self.heights[nr][nc] >= self.heights[row][col]:
                self.dfs(nr, nc, ocean)
        
        return

File: TypescriptSolutions/Neetcode150/graph_grid_problems/417_pacific-atlantic-water-flow.ts
export {};
const pacificAtlantic = (heights: number[][]): number[][] => {
    const ROWS = heights.length;
    const COLS = heights[0].length;
    const DIRS = [[0, 1], [1, 0], [-1, 0], [0, -1]];

    const pacific = new Set<string>();
    const atlantic = new Set<string>();

    const dfs = (row, col, ocean) => {
        if (ocean.has([row, col].toString())) return;

        ocean.add([row, col].toString());

        for (const [dx, dy] of DIRS) {
            const nr = row + dx;
            const nc = col + dy;

            if (nr < 0 || nc < 0 || nr > ROWS - 1 || nc > COLS - 1) continue;

            if (heights[nr][nc] >= heights[row][col]) {
                dfs(nr, nc, ocean);
            }
        }

        return;
    }

    for (let col = 0; col < COLS; col++) {
        dfs(0, col, pacific);
        dfs(ROWS - 1, col, atlantic);
    }

    for (let row = 0; row < ROWS; row++) {
        dfs(row, 0, pacific);
        dfs(row, COLS - 1, atlantic);
    }

    const res: number[][] = [];
    for (let row = 0; row < ROWS; row++) {
        for (let col = 0; col < COLS; col++) {
            if (pacific.has([row, col].toString()) && atlantic.has([row, col].toString())) {
                res.push([row, col]);
            }
        }
    }

    return res;
};